test/http-request.ts

```diff
@@ -0,0 +1,332 @@
+      assert.strictEqual(cloudReq.userAgent, 'some-agent');
+      assert.strictEqual(cloudReq.referer, 'some-referer');
+      assert.strictEqual(cloudReq.status, undefined);
+    });
+    it('should infer header values with caution', () => {
+      const req = {
+        method: 'GET',
+        url: 'http://google.com/',
+        headers: {
+          'user-agent': '',
+          referer: undefined,
+        },
+      } as http.IncomingMessage;
+      const cloudReq = makeHttpRequestData(req);
+      assert.strictEqual(cloudReq.protocol, 'http:');
+      assert.strictEqual(cloudReq.requestUrl, 'http://google.com/');
+      assert.strictEqual(cloudReq.requestMethod, 'GET');
+      assert.strictEqual(cloudReq.userAgent, undefined);
+      assert.strictEqual(cloudReq.referer, undefined);
+      assert.strictEqual(cloudReq.status, undefined);
+    });
+    it('should infer as many response values as possible', () => {
+      const RESPONSE_SIZE = 2048;
+      const req = {} as ServerRequest;
+      const res = {
+        statusCode: 200,
+        headers: {
+          'Content-Length': RESPONSE_SIZE,
+        },
+      } as unknown as http.ServerResponse;
+      res.getHeader = function () {
+        return 2048;
+      };
+      const cloudReq = makeHttpRequestData(req, res);
+      assert.strictEqual(cloudReq.status, 200);
+      assert.strictEqual(cloudReq.responseSize, RESPONSE_SIZE);
+    });
+    it('should convert latency to proto Duration', () => {
+      const fakeRequest = {headers: {}};
+      const fakeResponse = {};
+
+      const h1 = makeHttpRequestData(
+        fakeRequest as ServerRequest,
+        fakeResponse as ServerResponse,
+        1003
+      );
+      assert.deepStrictEqual(h1.latency, {seconds: 1, nanos: 3e6});
+
+      const h2 = makeHttpRequestData(
+        fakeRequest as ServerRequest,
+        fakeResponse as ServerResponse,
+        9003.1
+      );
+      assert.deepStrictEqual(h2.latency, {seconds: 9, nanos: 3.1e6});
+
+      // Make sure we nanos is uint32.
+      const h3 = makeHttpRequestData(
+        fakeRequest as ServerRequest,
+        fakeResponse as ServerResponse,
+        1.0000000001
+      );
+      assert.deepStrictEqual(h3.latency, {seconds: 0, nanos: 1e6});
+    });
+  });
+});
+
+describe('get trace and span from http-request', () => {
+  const FAKE_CONTEXT = {
+    // eslint-disable-next-line @typescript-eslint/no-unused-vars
+    extract: (headerWrapper: {}) => {},
+    generate: () => {},
+    // eslint-disable-next-line @typescript-eslint/no-unused-vars
+    inject: (headerWrapper: {}, spanContext: {}) => {},
+  };
+
+  const fakeContext = Object.assign({}, FAKE_CONTEXT);
+
+  const {getOrInjectTraceParent} = proxyquire('../src/http-request', {
+    '@opencensus/propagation-stackdriver': fakeContext,
+  });
+
+  describe('getTraceContext', () => {
+    it('should return an empty trace string when extraction fails and inject is false', () => {
```

Comment:
can you rephrase to be more clear? which extraction?

Reply:
rephrased!